RadixWallet/Clients/FactorSourcesClient/FactorSourcesClient+Live.swift

```diff
@@ -128,6 +132,64 @@ extension FactorSourcesClient: DependencyKey {
 			factorSourcesAsyncSequence: {
 				await profileStore.factorSourcesValues()
 			},
+			nextEntityIndexForFactorSource: { request in
+				let maybeNetworkID = request.networkID
+				let maybeFactorSourceID = request.factorSourceID
+				let mainBDFS = try await getMainDeviceFactorSource()
+				let factorSourceID = maybeFactorSourceID ?? mainBDFS.factorSourceID.embed()
+
+				let currentNetworkID = await getCurrentNetworkID()
+				let networkID = maybeNetworkID ?? currentNetworkID
+				let maybeNetwork: Profile.Network? = try? await profileStore.profile.network(id: networkID)
+
+				/// We CANNOT just use `entitiesControlledByFactorSource.count` since it is possible that
+				/// some users from Radix Babylon Wallet version 1.0.0 created accounts not sarting at
+				/// index `0` (since we had global indexing, shared by all FactorSources...), lets say that
+				/// only one account is controlled by a FactorSource `X`, having index `1`, then if we were
+				/// to used `entitiesControlledByFactorSource.count` for "next index" then that would be...
+				/// the value `1` AGAIN! Which does not work. Instead we need to read out the last path
+				/// component (index!) of the derivation paths of `entitiesControlledByFactorSource` and
+				/// find the MAX value and +1 on that. This also ensures that we are NOT "gap filling",
+				/// meaning that we do not want to use index `0` even if it was not used, where `1` was used, so
+				/// next index should be `2`, not `0` (which was free). The rationale is that it would just be
+				/// confusing and messy (for us not the least). Best to always increase. But it is important
+				/// to know  AccountRecoveryScan SHOULD find these "gap entities"!
+				func nextDerivationIndexForFactorSource(
+					entitiesControlledByFactorSource: some Collection<some EntityProtocol>
+				) -> HD.Path.Component.Child.Value {
+					let indicesOfEntitiesControlledByAccount = entitiesControlledByFactorSource
+						.compactMap { entity -> HD.Path.Component.Child.Value? in
+							switch entity.securityState {
+							case let .unsecured(unsecuredControl):
+								let factorInstance = unsecuredControl.transactionSigning
+								guard factorInstance.factorSourceID.embed() == factorSourceID else {
+									return nil
+								}
+								return factorInstance.derivationPath.index
+							}
+						}
+					guard !indicesOfEntitiesControlledByAccount.isEmpty else { return 0 }
+					let max = indicesOfEntitiesControlledByAccount.max()!
+					let nextIndex = max + 1
+					return nextIndex
+				}
+
+				if let network = maybeNetwork {
```

Comment:
Since we're never using maybeNetwork apart "non-unwrapped", I would personally just call it network. But of course I know you like long descriptive names.

Reply:
fixed